mqtt/mqtt_client.go

```diff
@@ -0,0 +1,184 @@
+package mqtt
+
+import (
+	"encoding/base64"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"github.com/btcsuite/btcd/btcec"
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	mqtt "github.com/eclipse/paho.mqtt.golang"
+	"log"
+	"math/big"
+	"os"
+	"time"
+
+	"crypto/rand"
+	"encoding/hex"
+)
+
+type mqttClient struct {
+	client        mqtt.Client
+	opts          *mqtt.ClientOptions
+	thePrivateKey []byte
+
+	// signedMsgPrefix is a special prefix that we'll prepend to any
+	// messages we sign/verify. We do this to ensure that we don't
+	// accidentally sign a sighash, or other sensitive material. By
+	// prepending this fragment, we mind message signing to our particular
+	// context.
+	signedMsgPrefix []byte
+}
+
+var CLIENT mqttClient
+
+var f mqtt.MessageHandler = func(client mqtt.Client, msg mqtt.Message) {
+	fmt.Printf("Messages published:\n  [topic= %s\n  payload= %s\n", msg.Topic(), msg.Payload())
+}
+
+func Init() {
+	//mqtt.DEBUG = log.New(os.Stdout, "[mqtt:debug]", 0)
+	mqtt.ERROR = log.New(os.Stdout, "[mqtt:error]", 0)
+	mqtt.CRITICAL = log.New(os.Stdout, "[mqtt:CRITICAL]", 0)
+	//mqtt.WARN = log.New(os.Stdout, "[mqtt:warn]", 0)
+
+	mqttScheme := getEnv("MQTT_SCHEME", "tcp")
+	mqttPort := getEnv("MQTT_PORT", "1883")
+	mqttHost := getEnv("MQTT_HOSTNAME", "host.docker.internal")
+	mqttClientID, _ := CLIENT.randomClientName()
+
+	password, mqttPublicKey, err := CLIENT.signTimestamp()
+	if err != nil {
+		panic(err)
+	}
+
+	fmt.Printf("password: %s mqttPublicKey: %s", password, mqttPublicKey)
+
+	brokerUri := fmt.Sprintf("%s://%s:%s", mqttScheme, mqttHost, mqttPort)
+
+	opts := mqtt.NewClientOptions()
+	opts.AddBroker(brokerUri)
+	opts.SetClientID(mqttClientID)
+
+	opts.SetUsername(mqttPublicKey)
+	opts.SetPassword(password)
+	opts.SetCleanSession(false)
+	opts.SetOnConnectHandler(func(c mqtt.Client) { CLIENT.connectedToLocalCallback(mqttPublicKey) })
+	opts.SetDefaultPublishHandler(f)
+
+	CLIENT.opts = opts
+	CLIENT.signedMsgPrefix = []byte("Lightning Signed Message:")
+
+	CLIENT.client = Client()
+}
+
+func Client() mqtt.Client {
+	client := mqtt.NewClient(CLIENT.opts)
+	if token := client.Connect(); token.Wait() && token.Error() != nil {
+		fmt.Println("=> connection error:", token.Error())
+	}
+	return client
+}
+
+func (c mqttClient) connectedToLocalCallback(pubkey string) {
+	c.client.Subscribe(pubkey+"/#", 0, func(c mqtt.Client, msg mqtt.Message) {
+		fmt.Printf("* [%s] %s\n", msg.Topic(), string(msg.Payload()))
+	})
+	pwd, pubkey, err := c.signTimestamp()
+	if err != nil {
+		log.Println("there was an error:", err)
+	}
+	c.client.Publish(pubkey+"/"+pwd, 0, false, "hi")
```

Comment:
So when I start up tribes I'm seeing a different pubkey and password logged that what we use to connect and see right before this error message 
```
password: Ymb_USC9ac_t-j90Qgpxp1KMJabeLc7rv0YJhjPrVIS27VpBzCwVgpsFOK0TpLdiE8ssq_yEBxQcIiihZjzMvjq0Of2_ mqttPublicKey: 029a99dd3c144f61d461f89bc8415fc81ff7144872c0c04e8657656c8cd007c10c[mqtt:error][client]   Failed to connect to a broker
```
are these supposed to be the same?

Reply:
nvm I see this is for the local callback ignore this